curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherIdentity.java

```diff
@@ -32,167 +38,152 @@
-            client.setData().forPath(PATH, new byte[]{});
-            timing.sleepABit();
-            assertEquals(1, watcher.count.get());
-        }
-        finally
-        {
-            CloseableUtils.closeQuietly(client);
-        }
-    }
-
-
-    @Test
-    public void testZKWatcher() throws Exception
-    {
-        Timing timing = new Timing();
-        CountZKWatcher watcher = new CountZKWatcher();
-        CuratorFramework client = CuratorFrameworkFactory.newClient(server.getConnectString(), timing.session(), timing.connection(), new RetryOneTime(1));
-        try
-        {
-            client.start();
-            client.create().forPath(PATH);
-            // Add twice the same watcher on the same path
-            client.getData().usingWatcher(watcher).forPath(PATH);
-            client.getData().usingWatcher(watcher).forPath(PATH);
-            // Ok, let's test it
-            client.setData().forPath(PATH, new byte[]{});
-            timing.sleepABit();
-            assertEquals(1, watcher.count.get());
-        }
-        finally
-        {
-            CloseableUtils.closeQuietly(client);
-        }
-    }
+public class TestWatcherIdentity extends BaseClassForTests {
+	private static final String PATH = "/foo";
+
+	private static class CountZKWatcher implements Watcher {
+		private final AtomicInteger count = new AtomicInteger(0);
+
+		@Override
+		public void process(WatchedEvent event) {
+			count.incrementAndGet();
+		}
+	}
+
+	@Test
+	public void testSameWatcherPerZKDocs() throws Exception {
+		CountZKWatcher actualWatcher = new CountZKWatcher();
+		Timing timing = new Timing();
+		CuratorFramework client = CuratorFrameworkFactory.newClient(server.getConnectString(), timing.session(),
+				timing.connection(), new RetryOneTime(1));
+		try {
+			client.start();
+			client.create().forPath("/test");
+
+			// per ZK docs, this watcher should only trigger once
+			client.checkExists().usingWatcher(actualWatcher).forPath("/test");
+			client.getData().usingWatcher(actualWatcher).forPath("/test");
+
+			client.setData().forPath("/test", "foo".getBytes());
+			client.delete().forPath("/test");
+			timing.sleepABit();
+			assertEquals(actualWatcher.count.getAndSet(0), 1);
+
+			client.create().forPath("/test");
+			client.checkExists().usingWatcher(actualWatcher).forPath("/test");
+			client.delete().forPath("/test");
+			timing.sleepABit();
+			assertEquals(actualWatcher.count.get(), 1);
+		} finally {
+			CloseableUtils.closeQuietly(client);
+		}
+	}
+
+	@Test
+	public void testSameCuratorWatcherPerZKDocs() throws Exception {
+		// Create variables for tracking behaviors of mock object
+		AtomicInteger actualWatcherCount = new AtomicInteger(0);
+		// Construct mock object
+		CuratorWatcher actualWatcher = mock(CuratorWatcher.class);
+		// Method Stubs
+		doAnswer((stubInvo) -> {
+			actualWatcherCount.incrementAndGet();
+			return null;
+		}).when(actualWatcher).process(any(WatchedEvent.class));
```

Comment:
We can further improve test by removing the method stub and the AtomicInteger variable. We can directly verify the behavior of `process(WacthedEvent)` by using Mockito.verify() with timeout function (so we can also remove `timing.sleepABit()`). This will make test condition more explict and elegant.